server/src/com/cloud/user/DomainManagerImpl.java

```diff
@@ -273,82 +289,133 @@ public boolean deleteDomain(long domainId, Boolean cleanup) {
-                    }
+                    checkDomainAccountsNetworksAndResourcesBeforeRemoving(domain);
+                }
 
-                    CloudRuntimeException e = new CloudRuntimeException("Can't delete the domain yet because it has " + msg);
-                    e.addProxyObject(domain.getUuid(), "domainId");
-                    throw e;
+                cleanupDomainOfferings(domain.getId());
+                CallContext.current().putContextParameter(Domain.class, domain.getUuid());
+                return true;
+            } catch (Exception ex) {
+                s_logger.error("Exception deleting domain with id " + domain.getId(), ex);
+                if (ex instanceof CloudRuntimeException)
+                    throw (CloudRuntimeException)ex;
+                else
+                    return false;
+            } finally {
+                //when success is false
+                if (rollBackState) {
+                    s_logger.debug("Changing domain id=" + domain.getId() + " state back to " + Domain.State.Active +
+                        " because it can't be removed due to resources referencing to it");
+                    domain.setState(Domain.State.Active);
+                    _domainDao.update(domain.getId(), domain);
                 }
             }
+        }
+        finally {
+            lock.unlock();
+            rollBackState = false;
+        }
+    }
 
-            cleanupDomainOfferings(domain.getId());
-            CallContext.current().putContextParameter(Domain.class, domain.getUuid());
-            return true;
-        } catch (Exception ex) {
-            s_logger.error("Exception deleting domain with id " + domain.getId(), ex);
-            if (ex instanceof CloudRuntimeException)
-                throw (CloudRuntimeException)ex;
-            else
-                return false;
-        } finally {
-            //when success is false
-            if (rollBackState) {
-                s_logger.debug("Changing domain id=" + domain.getId() + " state back to " + Domain.State.Active +
-                    " because it can't be removed due to resources referencing to it");
-                domain.setState(Domain.State.Active);
-                _domainDao.update(domain.getId(), domain);
-            }
+    /**
+     * Check domain resources before removing domain. There are 2 cases:
+     * <ol>
+     * <li>Domain doesn't have accounts for cleanup, non-removed networks, or dedicated resources</li>
+     * <ul><li>Delete domain</li></ul>
+     * <li>Domain has one of the following: accounts set for cleanup, non-removed networks, dedicated resources</li>
+     * <ul><li>Dont' delete domain</li><li>Fail operation</li></ul>
+     * </ol>
+     * @param domain domain to remove
+     * @throws CloudRuntimeException when case 2 or when domain cannot be deleted on case 1
+     */
+    protected void checkDomainAccountsNetworksAndResourcesBeforeRemoving(DomainVO domain) {
```

Comment:
I agree with you. I changed it to `removeDomainWithNoAccountsForCleanupNetworksOrDedicatedResources`. I think is quite more descriptive but still long